api/envoy/config/route/v3/route_components.proto

```diff
@@ -1645,6 +1646,25 @@ message RateLimit {
       string default_value = 3;
     }
 
+    // The following descriptor entry is appended when the
+    // :ref:`route entry metadata <envoy_api_field_config.route.v3.Route.metadata>` contains a key value:
+    //
+    // .. code-block:: cpp
+    //
+    //   ("<descriptor_key>", "<value_queried_from_route_entry_metadata>")
+    message RouteEntryMetaData {
```

Comment:
Did you consider just adding a type enum to the above `DynamicMetaData` message? The default can be what it does today and then another value can be route? WDYT?

Reply:
I did not consider that but could be a good idea. In my opinion then `DynamicMetaData` should be renamed to `MetaData` which is a breaking change. I am open to change the v4alpha part in this direction, but I would keep the current implementation for v3.

Reply:
We can handle the rename separately, but I think let's just go ahead and add the enum now? I think that will take us in the right direction.

/wait

Reply:
What about keeping DynamicMetaData as it is and adding a new MetaData action with the enum in both v3 and v4alpha instead of the RouteEntryMetaData action?
This allows the deprecation of DynamicMetadata in v4 and its removal it in v5 if I am understanding the API deprecation policy correctly.

Reply:
Sure that's fine!

/wait

Reply:
I changed the implementation, but I could not add the deprecation notice to v4alpha, the check_format insisted on removing it.
Can you help me what is the problem with it?

``` diff
diff --git a/api/envoy/config/route/v4alpha/route_components.proto b/api/envoy/config/route/v4alpha/route_components.proto
index 2d7af16d0..4c3f1e756 100644
--- a/api/envoy/config/route/v4alpha/route_components.proto
+++ b/api/envoy/config/route/v4alpha/route_components.proto
@@ -1582,6 +1582,9 @@ message RateLimit {
     // .. code-block:: cpp
     //
     //   ("<descriptor_key>", "<value_queried_from_dynamic_metadata>")
+    //
+    // .. attention::
+    //  This action has been deprecated in favor of the :ref:`envoy_api_msg_config.route.v4alpha.MetaData` action
     message DynamicMetaData {
       option (udpa.annotations.versioning).previous_message_type =
           "envoy.config.route.v3.RateLimit.Action.DynamicMetaData";
@@ -1652,7 +1655,10 @@ message RateLimit {
       HeaderValueMatch header_value_match = 6;
 
       // Rate limit on dynamic metadata.
-      DynamicMetaData dynamic_metadata = 7;
+      // .. attention::
+      //  This field has been deprecated in favor of the :ref:`envoy_api_field_config.route.v4alpha.Route.metadata` field
+      DynamicMetaData dynamic_metadata = 7
+          [deprecated = true, (envoy.annotations.disallowed_by_default) = true];
 
       // Rate limit on metadata.
       MetaData metadata = 8;
diff --git a/generated_api_shadow/envoy/config/route/v4alpha/route_components.proto b/generated_api_shadow/envoy/config/route/v4alpha/route_compo>
index 54152d16b..71533cbe5 100644
--- a/generated_api_shadow/envoy/config/route/v4alpha/route_components.proto
+++ b/generated_api_shadow/envoy/config/route/v4alpha/route_components.proto
@@ -1649,6 +1649,9 @@ message RateLimit {
     // .. code-block:: cpp
     //
     //   ("<descriptor_key>", "<value_queried_from_dynamic_metadata>")
+    //
+    // .. attention::
+    //  This action has been deprecated in favor of the :ref:`envoy_api_msg_config.route.v4alpha.MetaData` action
     message DynamicMetaData {
       option (udpa.annotations.versioning).previous_message_type =
           "envoy.config.route.v3.RateLimit.Action.DynamicMetaData";
@@ -1719,7 +1722,10 @@ message RateLimit {
       HeaderValueMatch header_value_match = 6;
 
       // Rate limit on dynamic metadata.
-      DynamicMetaData dynamic_metadata = 7;
+      // .. attention::
+      //  This field has been deprecated in favor of the :ref:`envoy_api_field_config.route.v4alpha.Route.metadata` field
+      DynamicMetaData dynamic_metadata = 7
+          [deprecated = true, (envoy.annotations.disallowed_by_default) = true];
 
       // Rate limit on metadata.
       MetaData metadata = 8;
```